File: sktmorph/engine/lakara.py
"""Lakāra normalization and classification."""

from typing import Optional, Tuple

# User-facing codes (constants.py) -> bundled SQLite codes
CANONICAL_TO_DB = {
    "plat": "plat",
    "plrt": "plrut",
    "plot": "plot",
    "plan": "plang",
    "pvidhilin": "pvidhiling",
    "alat": "alat",
    "alrt": "alrut",
    "alot": "alot",
    "alan": "alang",
    "aling": "aashirling",
    "alit": "alit",
}

DB_TO_CANONICAL = {v: k for k, v in CANONICAL_TO_DB.items()}
DB_TO_CANONICAL["pashirling"] = "pvidhilin"
DB_TO_CANONICAL["plit"] = "plat"  # parasmaipada perfect uses distinct endings

# Lakāra family drives stem formation and ending tables
LAKARA_FAMILIES = {
    "plat": "lat",
    "alat": "lat",
    "plot": "lot",
    "alot": "lot",
    "plrut": "lrt",
    "alrut": "lrt",
    "plang": "lang",
    "alang": "lang",
    "pvidhiling": "vidhilin",
    "pashirling": "vidhilin",
    "aashirling": "vidhilin",
    "avidhiling": "vidhilin",
    "plit": "lit",
    "alit": "lit",
}


def normalize_lakara(lakara: str) -> Tuple[str, str]:
    """Return (canonical_code, db_code)."""
    code = lakara.strip()
    if code in CANONICAL_TO_DB:
        return code, CANONICAL_TO_DB[code]
    if code in DB_TO_CANONICAL:
        return DB_TO_CANONICAL[code], code
    return code, code


def lakara_family(db_lakara: str) -> Optional[str]:
    return LAKARA_FAMILIES.get(db_lakara)


def pada_from_lakara(db_lakara: str) -> str:
    if db_lakara == "plit":
        return "P"
    if db_lakara.startswith("a"):
        return "A"
    return "P"


def resolve_pada(db_lakara: str, root_pada: str) -> Optional[str]:
    """Effective pada for this lakāra given root pada restriction."""
    if db_lakara == "plit":
        return "P"
    if db_lakara.startswith("a"):
        return "A"
    if db_lakara.startswith("p"):
        return "P"
    return None


def kartari_compatible(root_pada: str, lakara: str) -> bool:
    """Whether tinanta engine should produce forms for this root/lakāra pair."""
    _, db_lk = normalize_lakara(lakara)
    if db_lk.startswith("a") or db_lk == "alit":
        return root_pada in ("A", "U")
    if db_lk == "plit":
        return root_pada in ("P", "U")
    if root_pada == "A":
        return False
    if root_pada == "P" and db_lk.startswith("a"):
        return False
    return True

File: src/engine/lakara.rs
//! Port of engine/lakara.py

//! =============================================================================
//! src/engine/lakara.rs: Pāṇini/Kaumudī implementation — extreme commenting pass (2026-09-01)
//! ---------------------------------------------------------------------------
//! Purpose: see inline block comments below. Every public/private block is
//! documented with sūtra reference, input/output, and edge-case notes.
//! Script: SLP1 internally; Devanagari only at demo boundary.
//! Flow: dhātu → it-strip → aṅga/vikaraṇa → lakāra/ending → sandhi → surface.
//! Gold DB is cross-check only, never source of truth.
//! =============================================================================
pub fn normalize_lakara(lakara: &str) -> (String, String) {
    let code = lakara.trim();
    // canonical -> db
    let map: &[(&str,&str)] = &[
        ("plat","plat"),("plrt","plrut"),("plot","plot"),("plan","plang"),("pvidhilin","pvidhiling"),("plit","plit"),
        ("plun","plun"),("pashirling","pashirling"),
        ("alat","alat"),("alrt","alrut"),("alot","alot"),("alan","alang"),("avidhilin","avidhiling"),("aling","aashirling"),("alit","alit"),
        ("alun","alun"),
    ];
    // — for — iterate dhātu/ending variants; sūtra gating, see comments above.
    for (c,db) in map { if *c==code { return (c.to_string(), db.to_string()); } }
    // db -> canonical
    let rev: &[(&str,&str)] = &[
        ("plat","plat"),("plrut","plrt"),("plot","plot"),("plang","plan"),("pvidhiling","pvidhilin"),
        ("alat","alat"),("alrut","alrt"),("alot","alot"),("alang","alan"),("avidhiling","avidhilin"),("aashirling","aling"),("alit","alit"),
        ("plit","plit"),("plun","plun"),("alun","alun"),("pashirling","pashirling"),
    ];
    // — for — iterate dhātu/ending variants; sūtra gating, see comments above.
    for (db,c) in rev { if *db==code { return (c.to_string(), db.to_string()); } }
    (code.to_string(), code.to_string())
}

// ---------------------------------------------------------------------------
// fn `lakara_family`: purpose, inputs→outputs, edge cases.
// Pāṇini step; see Kaumudī ordering. SLP1 I/O. No DB fallback.
// ---------------------------------------------------------------------------
pub fn lakara_family(db_lakara: &str) -> Option<String> {
    // — match — pada/lakāra/gaṇa dispatch; sūtra gating, see comments above.
    match db_lakara {
        "plat"|"alat" => Some("lat".into()),
        "plot"|"alot" => Some("lot".into()),
        "plrut"|"alrut" => Some("lrt".into()),
        "plang"|"alang" => Some("lang".into()),
        "pvidhiling"|"avidhiling" => Some("vidhilin".into()),
        "plit"|"alit" => Some("lit".into()),
        "plun"|"alun" => Some("lun".into()),
        "pashirling"|"aashirling" => Some("ashir".into()),
        _ => None,
    }
}

// ---------------------------------------------------------------------------
// fn `pada_from_lakara`: purpose, inputs→outputs, edge cases.
// Pāṇini step; see Kaumudī ordering. SLP1 I/O. No DB fallback.
// ---------------------------------------------------------------------------
pub fn pada_from_lakara(db_lakara: &str) -> String {
    // — if-branch — condition → aṅga/sandhi step; sūtra gating, see comments above.
    if db_lakara=="plit" { return "P".into(); }
    // — if-branch — condition → aṅga/sandhi step; sūtra gating, see comments above.
    if db_lakara.starts_with('a') { return "A".into(); }
    "P".into()
}

// ---------------------------------------------------------------------------
// fn `kartari_compatible`: purpose, inputs→outputs, edge cases.
// Pāṇini step; see Kaumudī ordering. SLP1 I/O. No DB fallback.
// ---------------------------------------------------------------------------
pub fn kartari_compatible(root_pada: &str, lakara: &str) -> bool {
    let (_, db) = normalize_lakara(lakara);
    // — if-branch — condition → aṅga/sandhi step; sūtra gating, see comments above.
    if db.starts_with('a') || db=="alit" { return matches!(root_pada, "A"|"U"); }
    // — if-branch — condition → aṅga/sandhi step; sūtra gating, see comments above.
    if db=="plit" { return matches!(root_pada, "P"|"U"); }
    // — if-branch — condition → aṅga/sandhi step; sūtra gating, see comments above.
    if root_pada=="A" { return false; }
    true
}
